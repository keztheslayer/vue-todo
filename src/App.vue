<template>
    <div class="container">
        <div class="row">
            <div class="col-lg-8 col-12 mx-auto">
                <div class="todo">
                    <h1 class="todo__title">
                        {{ title }}
                    </h1>
                    <transition-group
                        name="slide"
                        tag="div"
                        mode="in-out"
                    >
                        <app-item
                            v-for="(item, index) in tasksList"
                            :key="`item ${index}`"
                            :index="index"
                            :text="item.text"
                            :is-checked="item.isChecked"
                            :hide-divider="index < tasksList.length - 1"
                        />
                    </transition-group>
                    <app-form />
                    <hr>
                    <app-footer />
                </div>
            </div>
        </div>
    </div>
</template>

<script>
import AppFooter from './components/Footer.vue';
import AppForm from './components/Form.vue';
import AppItem from './components/Item.vue';
import { mapGetters } from 'vuex';

const STORAGE_KEY = 'todo-storage';

export default {
    components : {
        AppFooter,
        AppForm,
        AppItem,
    },
    computed : {
        ...mapGetters( [
            'tasksList',
            'title',
            'finishedTasks',
        ] ),
    },
    updated() {
        this.saveToLocal();
    },
    methods : {
        saveToLocal() {
            localStorage.setItem( STORAGE_KEY, JSON.stringify( this.tasksList ) );
        },
    },
};
</script>

<style lang="less">
    body {
        background-image: linear-gradient( 135deg, #FFD26F 10%, #3677FF 100%);
        min-height: 100vh;
    }
    .todo {
        background: #fff;
        box-shadow: 0px 3px 3px rgba(17, 56, 86, 0.05);
        margin-top: 24px;
        border: 1px solid #eee;
        padding: 24px;
        border-radius: 10px;

        &__title {
            font-size: 36px;
            color: #333;
            margin-bottom: 24px;

            @media (max-width: 768px){
                font-size: 26px;
            }
        }
    }

    .slide-enter-active, .slide-leave-active {
        transition: all 0.3s;
    }

    .slide-enter {
        opacity: 0;
        transform: translateY(30px);
    }

    .slide-leave-to {
        opacity: 0;
    }

    .slide-move {
        transition: 0.3s;
    }

    @import "https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css";
</style>
