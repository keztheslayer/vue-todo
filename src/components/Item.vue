<template>
    <div
        class="todo__list-item"
        :class="isChecked ? 'todo__list-item_state_finished' : ''"
    >
        <div class="todo__list-item-wrapper">
            <label class="todo__label">
                <input
                    v-model="isChecked"
                    class="todo__checkbox"
                    type="checkbox"
                    @click="checkItem( index )"
                >
                <span class="todo__list-item-text">{{ text }}</span>
            </label>
            <button
                class="todo__list-item-edit"
                @click="editTask( index )"
            />
            <button
                type="button"
                class="close todo__remove-btn"
                @click="removeTask( index )"
            >
                <span>&times;</span>
            </button>
        </div>
        <hr v-if="hideDivider">
    </div>
</template>

<script>
import { mapActions } from 'vuex';

export default {
    props : {
        isChecked : {
            type    : Boolean,
            default : false,
        },
        text : {
            type    : String,
            default : '',
        },
        index : {
            type    : Number,
            default : 0,
        },
        hideDivider : {
            type    : Boolean,
            default : true,
        },
    },
    methods : {
        ...mapActions( [
            'editTask',
            'checkItem',
            'removeTask',
        ] ),
    },
};
</script>

<style lang="less">
.todo {
    &__label {
        cursor: pointer;
        color: #333;
        font-size: 18px;
        margin: 0;
        display: block;
        flex-grow: 1;
    }

    &__list-item {
        position: relative;
        margin-bottom: 6px;

        &:last-child {
            margin-bottom: 0;
        }
    }

    &__list-item-text {
        transition: .15s ease-in;
    }

    &__list-item-wrapper {
        display: flex;
        flex-wrap: nowrap;
        justify-content: space-between;
        align-items: center;
    }

    &__list-item-edit {
        display: block;
        width: 24px;
        height: 24px;
        background: transparent;
        border: none;
        appearance: none;
        margin-right: 12px;

        &:focus {
            outline: none;
        }
    }

    &__list-item-edit:before {
        content: '\f040';
        cursor: pointer;
        color: #000;
        font: 15px "fontAwesome";
        line-height: 1;
        width: 1em;
        display: inline-block;
        opacity: 0.5;
        transition: opacity 0.3s;
    }
    &__list-item-edit:hover:before {
        opacity: 1;
    }

    &__checkbox + &__list-item-text:hover:before {
        color: #fe4365;
    }

    &__checkbox {
        display: none;
    }

    &__checkbox + &__list-item-text:before {
        content: "";
        color: #dddfe6;
        font-family: "fontAwesome";
        line-height: 1;
        width: 1em;
        display: inline-block;
        margin-right: 8px;
    }

    &__checkbox:checked + &__list-item-text:before {
        content: "";
        color: #fe4365;
        animation: tick 150ms ease-in;
    }

    &__checkbox:checked + &__list-item-text {
        color: #7e7e7e;
    }

    &__checkbox:disabled + &__list-item-text:before {
        content: "";
        color: #dddfe6;
    }

    &__remove-btn {
        display: block;
        width: 24px;
        height: 24px;

        &:focus {
            outline: none;
        }
    }
}

@keyframes tick {
    0% {
        transform: scale(0);
    }
    90% {
        transform: scale(1.3);
    }
    100% {
        transform: scale(1);
    }
}
</style>

