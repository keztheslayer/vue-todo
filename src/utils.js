export const STORAGE_KEY = 'todo-storage';
