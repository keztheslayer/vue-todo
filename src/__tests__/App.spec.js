import App from '../App.vue';
import Item from '../components/Item.vue';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import { testTasks, STORAGE_KEY } from '../utils';

const localVue = createLocalVue();

localVue.use( Vuex );

describe( 'App.vue', () => {
    let getters;

    let store;

    let state;

    beforeEach( () => {
        state = {
            tasks : testTasks,
            title : 'Example app title',
        };

        getters = {
            tasksList() {
                return state.tasks;
            },
            title() {
                return state.title;
            },
            totalItems() {
                return state.tasks.length;
            },
            finishedTasks() {
                return state.tasks.filter( item => item.isChecked ).length;
            },
        };

        store = new Vuex.Store( {
            state,
            getters,
        } );
    } );

    describe( 'Correct visual rendering', () => {
        it( 'Displays one Item for 1 task in store', () => {
            const wrapper = shallowMount(
                App,
                {
                    store,
                    localVue,
                }
            );

            expect( wrapper.findAll( Item ) ).toHaveLength( getters.totalItems() );
        } );

        it( 'Renders correct markup', () => {
            const wrapper = shallowMount(
                App,
                {
                    store,
                    localVue,
                }
            );

            expect( wrapper.element ).toMatchSnapshot();
        } );
    } );

    describe( 'works with localStorage', () => {
        it( 'saveToLocal() method saves tasks list to localStorage', () => {
            const wrapper = shallowMount(
                App,
                {
                    store,
                    localVue,
                }
            );

            wrapper.vm.saveToLocal();

            const LOCAL_LIST = JSON.parse( localStorage.getItem( STORAGE_KEY ) );

            expect( LOCAL_LIST ).toEqual( getters.tasksList() );
        } );

        it( 'calls saveToLocal after an app-update hook', async () => {
            const saveToLocal = jest.fn();
            const wrapper = shallowMount(
                App,
                {
                    store,
                    localVue,
                    methods : {
                        saveToLocal,
                    },
                }
            );

            wrapper.vm.$forceUpdate();

            await Vue.nextTick();

            expect( saveToLocal ).toHaveBeenCalled();
        } );
    } );
} );
