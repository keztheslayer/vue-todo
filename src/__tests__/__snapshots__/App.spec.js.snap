// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue renders correct markup 1`] = `
<div
  class="container"
>
  <div
    class="row"
  >
    <div
      class="col-lg-8 col-12 mx-auto"
    >
      <div
        class="todo"
      >
        <h1
          class="todo__title"
        >
          
                    Example app title
                
        </h1>
         
        <transition-group-stub
          mode="in-out"
          name="slide"
          tag="div"
        >
          <app-item-stub
            hidedivider="true"
            index="0"
            text="Task 1"
          />
          <app-item-stub
            hidedivider="true"
            index="1"
            text="Task 2"
          />
          <app-item-stub
            index="2"
            ischecked="true"
            text="Task 3"
          />
        </transition-group-stub>
         
        <app-form-stub />
         
        <hr />
         
        <app-footer-stub />
      </div>
    </div>
  </div>
</div>
`;
